src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +29,119 @@ class DictionarySegment : public BaseSegment {
+  T get(const size_t chunk_offset) const { return value_by_value_id(ValueID{_attribute_vector->get(chunk_offset)}); }
 
   // dictionary segments are immutable
-  void append(const AllTypeVariant& val) override;
+  void append(const AllTypeVariant& val) {
+    throw std::runtime_error("Dictionary segments are immutable. You shall not append anything.");
+  }
 
   // returns an underlying dictionary
-  std::shared_ptr<const std::vector<T>> dictionary() const;
+  std::shared_ptr<const std::vector<T>> dictionary() { return _dictionary; }
 
   // returns an underlying data structure
-  std::shared_ptr<const BaseAttributeVector> attribute_vector() const;
-
-  // return the value represented by a given ValueID
-  const T& value_by_value_id(ValueID value_id) const;
+  std::shared_ptr<BaseAttributeVector> attribute_vector() const { return _attribute_vector; }
 
   // returns the first value ID that refers to a value >= the search value
   // returns INVALID_VALUE_ID if all values are smaller than the search value
-  ValueID lower_bound(T value) const;
+  ValueID lower_bound(T value) const {
+    auto dictionary_size = _dictionary->size();
+    for (ValueID dictionary_index = ValueID{0}; dictionary_index < dictionary_size; ++dictionary_index) {
+      if (value_by_value_id(dictionary_index) >= value) {
+        return dictionary_index;
+      }
+    }
+
+    return INVALID_VALUE_ID;
+  }
 
   // same as lower_bound(T), but accepts an AllTypeVariant
-  ValueID lower_bound(const AllTypeVariant& value) const;
+  ValueID lower_bound(const AllTypeVariant& value) const { return lower_bound(static_cast<T>(value)); }
 
   // returns the first value ID that refers to a value > the search value
   // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
-  ValueID upper_bound(T value) const;
+  ValueID upper_bound(T value) const {
+    auto dictionary_size = _dictionary->size();
+    for (ValueID dictionary_index = ValueID{0}; dictionary_index < dictionary_size; ++dictionary_index) {
+      if (value_by_value_id(dictionary_index) > value) {
+        return dictionary_index;
+      }
+    }
+
+    return INVALID_VALUE_ID;
+  }
 
   // same as upper_bound(T), but accepts an AllTypeVariant
-  ValueID upper_bound(const AllTypeVariant& value) const;
+  ValueID upper_bound(const AllTypeVariant& value) const { return upper_bound(static_cast<T>(value)); }
 
-  // return the number of unique_values (dictionary entries)
-  size_t unique_values_count() const;
+  // return the number of _dictionary (dictionary entries)
+  size_t unique_values_count() const { return _dictionary->size(); }
 
   // return the number of entries
-  ChunkOffset size() const override;
+  ChunkOffset size() const { return _attribute_vector->size(); }
 
   // returns the calculated memory usage
-  size_t estimate_memory_usage() const final;
+  size_t estimate_memory_usage() const final {
+    auto attributeVecMem = _attribute_vector->width() * _attribute_vector->size();
```

Comment:
Please follow the contribution guidelines: `Variables: lowercase separated by underscores, e.g., chunk_size.`